Initialise union of static lifetime with zeros

tautschnig opened the pull request:
The C standard does not specify the value of bytes that do not belong to
the object representation of a member referred to in an initializer
list, but compilers seem to ensure those bytes are zero (possibly by
using the .bss section).

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Initialise union of static lifetime with zeros

The C standard does not specify the value of bytes that do not belong to
the object representation of a member referred to in an initializer
list, but compilers seem to ensure those bytes are zero (possibly by
using the .bss section).

kroening commented:
It would make sense to have a warning in the manual about this one as well.

tautschnig commented:
Marking do-not-merge as this requires fixes in dump-c (or some approach not involving `byte_update`).

kroening commented:
Looking at this, I am wondering whether or not we might want to have a 'strict mode', which would not make compiler or architecture-specific assumptions.

tautschnig commented:
> Looking at this, I am wondering whether or not we might want to have a 'strict mode', which would not make compiler or architecture-specific assumptions.

Well, at least we should make sure we can configure all those architecture/compiler-specific assumptions. We have quite a lot of this in `configt` already, but handling of unions is not among that yet. I'll work on adding that union-configuration part, but I'd be happy to hear about more aspects that come to mind.

tautschnig committed:
dump-c regression test of union initializers

Even in upcoming changes there must not be a situation where such
initializers cannot be dumped in a compilable fashion.

tautschnig committed:
Bitvector decoding of pointer-typed constants must use bvrep

The regression test demonstrated that decoding non-null pointer-typed
constants were not correctly decoded to their integer representation.

tautschnig committed:
dump-c: rewrite byte_updates of unions to union expressions

C does not have byte_update expressions.

tautschnig committed:
Simplify byte updates of constants

Mirror what we already did for byte extract operations.

tautschnig committed:
Simplify initializer to remove byte_update

We introduce byte updates to soundly prepare union initializers, but
such byte updates are not part of the C standard and thus aren't
expected by dump-c. Run simplify to clean them up, when possible.

tautschnig committed:
bits2expr: create union expressions when components aren't full width

A union_exprt can refer to components of a type that is smaller than the
full bit width of the union. Consequently, bits2expr should accept bit
strings for such (smaller) members.

tautschnig committed:
expr2bits: bit strings for null pointers must have pointer width

The order of constructor arguments was wrong, resulting in 48 copies
(the ASCII index of '0') instead of bit-width copies, being used.

tautschnig committed:
C front-end: clean up use of ID_designated_initializer

The parser unnecessarily introduced an ID_designated_initializer into
offsetof when no id was required, thereby creating an expression with
ID_designated_initializer that had a different shape from all other uses
of ID_designated_initializer. With this cleaned up, enable expr2c to
print designated-initializer expressions.

tautschnig commented:
Trying to make sure we can successfully dump union expressions after fixing zero initialisation via byte updates became quite the bug squashing party. I started out by creating a test for dump-c, which I then noticed CBMC would not produce the correct verification result for. Here is the history as it may help to understand the flood of commits:

1. Added a test that dump-c ought to reproduce correctly (42b4740).
2. Figured out that we could also check for "VERIFICATION SUCCESSFUL" as part of this test, but that required a bitvector flattening bugfix (completely unrelated to to dump-c, the original source would not yield the correct verification result) (7e02b53).
3. Wanted to finally add support to dump-c (eventually done in db4f25c), but that required working `ID_designated_initializer` output. Thus fixed that first (a835405).
4. Now it is possible to use `byte_update` in initialising unions without breaking dump-c's output thereof (a8592a7).
5. Wanted to use `simplify_expr` to resolve this `byte_update` in common cases. This required extending the simplifier (9a51a92) and fixing bugs in `bits2expr`/`expr2bits` (e8ac4d6, ed75598).
6. Finally the simplifier can be used successfully for the desired `byte_update` removal (ac31213).

peterschrammel commented on src/util/simplify_utils.cpp:
```diff
@@ -410,7 +417,7 @@ expr2bits(const exprt &expr, bool little_endian, const namespacet &ns)
     {
       if(value == ID_NULL && config.ansi_c.NULL_is_zero)
-        return std::string('0', to_bitvector_type(type).get_width());
+        return std::string(to_bitvector_type(type).get_width(), '0');
```
Good catch!
